### PR TITLE
Demo for a complex interface : position, velocity, effort, gain (Kp, Kd)

### PR DESCRIPTION
**Pull request by olivier-stasse**
**Objective**
This pull request demonstrates a quadruped robot with a complex interface: position, velocity, effort, gain Kp and Kd.

It is based on the Solo robot described [here](https://github.com/open-dynamic-robot-initiative/robot_properties_solo/tree/master/src/robot_properties_solo/robot_properties_solo)

This robot has some feature which are currently specific to its power electronics and the code implemented there.
Namely changing on line the gains was not possible in ros-control.

**Interface**
The  state and control interfaces are gathered in a std::set<<std::string>> inside the `rrbot_demo_hardware::quadruped` namespace to avoid clashes with other possible examples.
It simplifies the verification between the model ( in `ros2_demo_robot/description/rrbot/rrbot_system_quadruped.urdf.xacro`) and what the plugin is looking for.

**Status**
The current pull-request is set as a draft for comments.

Do you want this example ?

Possibly the location of the robot model is not very good and should be seperated. They are references to visual meshes which are not included. Should we include the meshes or remove their references in the xacro file ?

### CODE REVIEW

File: ros2_control_demo_hardware/src/rrbot_system_quadruped.cpp [hunk 1]
```diff
@@ -0,0 +1,320 @@
...
+}
+
+}  // namespace quadruped
+
+}  // namespace ros2_control_demo_hardware
+
+#include "pluginlib/class_list_macros.hpp"
+
+PLUGINLIB_EXPORT_CLASS(
+  ros2_control_demo_hardware::quadruped::RRBotSystemQuadrupedHardware,
```
Comment (destogl) [inline ros2_control_demo_hardware/src/rrbot_system_quadruped.cpp:318]: There is no need to introduce an additional namespace. Or are you planning to have multiple examples for this category?

```suggestion
  ros2_control_demo_hardware::RRBotSystemQuadrupedHardware,
```

File: ros2_control_demo_robot/description/rrbot/rrbot_quadruped.urdf.xacro [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+  <!-- This file is based on: https://atlas.is.localnet/confluence/display/AMDW/Quadruped+URDF+Files -->
+  <link name="base_link">
+    
+    <!-- BASE LINK INERTIAL -->
+    <inertial>
+        <origin xyz="0 0 0" rpy="0 0 0" />
+        <mass value="1.43315091"/>
+        <!-- The base is extremely symmetrical. -->
+        <inertia ixx="0.00578574" ixy="0.0" ixz="0.0"
+                                  iyy="0.01938108" iyz="0.0"
```
Comment (destogl) [inline ros2_control_demo_robot/description/rrbot/rrbot_quadruped.urdf.xacro:26]: alignment is strange here.

File: ros2_control_demo_robot/description/rrbot/rrbot_system_quadruped.ros2_control.xacro [hunk 3]
```diff
@@ -0,0 +1,83 @@
+<?xml version="1.0"?>
+<robot xmlns:xacro="http://www.ros.org/wiki/xacro">
+
+  <xacro:macro name="rrbot_system_quadruped_actuator" params="name">
```
Comment (destogl) [inline ros2_control_demo_robot/description/rrbot/rrbot_system_quadruped.ros2_control.xacro:4]: very nice use of macros!

File: ros2_control_demo_robot/description/rrbot_system_quadruped.urdf.xacro [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+Copied and modified from https://github.com/open-dynamic-robot-initiative/robot_properties_solo
+-->
+<robot xmlns:xacro="http://www.ros.org/wiki/xacro" name="quadruped_robot">
+
+  <!-- Import RRBot macro -->
+  <xacro:include filename="rrbot/rrbot_quadruped.urdf.xacro" />
+
+  <!-- Import RRBot ros2_control description -->
+  <xacro:include filename="rrbot/rrbot_system_quadruped.ros2_control.xacro" />
+  <!-- <xacro:include filename="$(find ros2_control_demo_robot)/description/rrbot/rrbot_system_position_only.ros2_control.xacro" /> -->
```
Comment (destogl) [inline ros2_control_demo_robot/description/rrbot_system_quadruped.urdf.xacro:13]: please use this line if it's working

File: ros2_control_demo_robot/description/rrbot_system_quadruped.urdf.xacro [hunk 5]
```diff
@@ -0,0 +1,24 @@
+<?xml version="1.0"?>
+<!-- Quadruped robot  -->
+<!--
+Copied and modified from https://github.com/open-dynamic-robot-initiative/robot_properties_solo
+-->
+<robot xmlns:xacro="http://www.ros.org/wiki/xacro" name="quadruped_robot">
+
+  <!-- Import RRBot macro -->
+  <xacro:include filename="rrbot/rrbot_quadruped.urdf.xacro" />
```
Comment (destogl) [inline ros2_control_demo_robot/description/rrbot_system_quadruped.urdf.xacro:9]: use find command as in the commented code below

File: ros2_control_demo_robot/launch/rrbot_system_quadruped.launch.py [hunk 6]
```diff
@@ -0,0 +1,52 @@
+# Copyright 2020 ROS2-Control Development Team (2020)
```
Comment (destogl) [inline ros2_control_demo_robot/launch/rrbot_system_quadruped.launch.py:1]:
```suggestion
# Copyright 2021, ros2_control Development Team
```

**Review by destogl — COMMENTED** [review]
this looks very good! Are there any controllers missing you need for this demo?

File: ros2_control_demo_hardware/CMakeLists.txt [hunk 7]
```diff
@@ -1,6 +1,8 @@
 cmake_minimum_required(VERSION 3.5)
 project(ros2_control_demo_hardware)
 
+set(CMAKE_VERBOSE_MAKEFILE TRUE)
```
[the remark this answers is not in the document]
Comment (olivier-stasse) [reply ros2_control_demo_hardware/CMakeLists.txt]: Fixed in fc203dc

File: ros2_control_demo_hardware/include/ros2_control_demo_hardware/rrbot_system_quadruped.hpp [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+#ifndef ROS2_CONTROL_DEMO_HARDWARE__RRBOT_SYSTEM_QUADRUPED_HPP_
+#define ROS2_CONTROL_DEMO_HARDWARE__RRBOT_SYSTEM_QUADRUPED_HPP_
+
+#include <memory>
+#include <string>
+#include <vector>
+#include <set>
+
+#include "rclcpp/macros.hpp"
+
```
[the remark this answers is not in the document]
Comment (olivier-stasse) [reply ros2_control_demo_hardware/include/ros2_control_demo_hardware/rrbot_system_quadruped.hpp]: Fix in b1446c5

File: ros2_control_demo_hardware/include/ros2_control_demo_hardware/rrbot_system_quadruped.hpp [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+// limitations under the License.
+
+
+#ifndef ROS2_CONTROL_DEMO_HARDWARE__RRBOT_SYSTEM_QUADRUPED_HPP_
+#define ROS2_CONTROL_DEMO_HARDWARE__RRBOT_SYSTEM_QUADRUPED_HPP_
+
+#include <memory>
+#include <string>
+#include <vector>
+#include <set>
```
[the remark this answers is not in the document]
Comment (olivier-stasse) [reply ros2_control_demo_hardware/include/ros2_control_demo_hardware/rrbot_system_quadruped.hpp]: Fix in b1446c5

File: ros2_control_demo_hardware/include/ros2_control_demo_hardware/rrbot_system_quadruped.hpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (olivier-stasse) [reply ros2_control_demo_hardware/include/ros2_control_demo_hardware/rrbot_system_quadruped.hpp]: Is it not needed to use RCLCPP_SHARED_PTR_DEFINITIONS ?

File: ros2_control_demo_hardware/src/rrbot_system_quadruped.cpp [hunk 10]
```diff
@@ -0,0 +1,320 @@
...
+  hw_start_sec_ = stod(info_.hardware_parameters["example_param_hw_start_duration_sec"]);
+  hw_stop_sec_ = stod(info_.hardware_parameters["example_param_hw_stop_duration_sec"]);
+  hw_slowdown_ = stod(info_.hardware_parameters["example_param_hw_slowdown"]);
+  hw_states_.resize(
+    info_.joints.size(),
+    {std::numeric_limits<double>::quiet_NaN(),
+      std::numeric_limits<double>::quiet_NaN(),
+      std::numeric_limits<double>::quiet_NaN(),
+      std::numeric_limits<double>::quiet_NaN(),
+      std::numeric_limits<double>::quiet_NaN()});
```
[the remark this answers is not in the document]
Comment (olivier-stasse) [reply ros2_control_demo_hardware/src/rrbot_system_quadruped.cpp:49]: I tried, but I think this is the default value for each joint. Each of them has 5 variables.

[the remark this answers is not in the document]
Comment (olivier-stasse) [reply ros2_control_demo_hardware/src/rrbot_system_quadruped.cpp:49]: But may be I misunderstood what should be in each states.

**Comment by olivier-stasse** [comment]
# Current status
Implements `accept_command_resource_claim`
Uses `forward_command_controller/MultipleInterfaceForwardController`

# Needs 
https://github.com/ros-controls/ros2_control/pull/322
https://github.com/ros-controls/ros2_controllers/pull/154

# To perform the tests
Can perform separetly `forward_command_controller_position` and `forward_command_controller_pos_vel_eff_gain`.
Switch is not working.
Read the README.md 

# Current problem
When trying to switch:
```[controller_manager]: Resource conflict for controller 'forward_command_controller_position'. Command interface 'FL_HFE/position' is already claimed.```

**Pull request closed without merging**